torchvision/datasets/mnist.py

```diff
@@ -286,25 +288,175 @@ def download(self):
+            what = 'train' if train else 'test'
+        if not self.subsets.get(what):
+            raise RuntimeError("Argument 'what' should be one of: \n  " +
+                               repr(tuple(self.subsets.keys())) )
+        self.what = what
+        self.compat = compat
+        self.data_file = what + '.pt'
+        self.training_file = self.data_file
+        self.test_file = self.data_file
+        super(QMNIST, self).__init__(root, train, **kwargs)
+
+    def download(self):
+        """Download the QMNIST data if it doesn't exist in processed_folder already.
+           Note that we only download what has been asked for (argument 'what').
+        """
+        if self._check_exists():
+            return
+        makedir_exist_ok(self.raw_folder)
+        makedir_exist_ok(self.processed_folder)
+        urls = self.urls[self.subsets[self.what]]
+        files = []
+
+        # download data files if not already there
+        for url in urls:
+            filename = url.rpartition('/')[2]
+            file_path = os.path.join(self.raw_folder, filename)
+            if not os.path.isfile(file_path):
+                download_url(url, root=self.raw_folder, filename=filename, md5=None)
+            files.append(file_path)
+
+        # process and save as torch files
+        print('Processing...')
+        data = read_sn3_pascalvincent_tensor(files[0])
+        assert(data.dtype == torch.uint8)
+        assert(data.ndimension() == 3)
+        targets = read_sn3_pascalvincent_tensor(files[1]).long()
+        assert(targets.ndimension() == 2)
+        if self.what == 'test10k':
+            data = data[0:10000,:,:].clone()
+            targets = targets[0:10000,:].clone()
+        if self.what == 'test50k':
+            data = data[10000:,:,:].clone()
+            targets = targets[10000:,:].clone()
+        with open(os.path.join(self.processed_folder, self.data_file), 'wb') as f:
+            torch.save((data, targets), f)
+
+    def __getitem__(self, index):
+        # redefined to handle the compat flag
+        img, target = self.data[index], self.targets[index]
+        img = Image.fromarray(img.numpy(), mode='L')
+        if self.transform is not None:
+            img = self.transform(img)
+        if self.compat:
+            target = int(target[0])
+        if self.target_transform is not None:
+            target = self.target_transform(target)
+        return img, target
+
+    def extra_repr(self):
+        return "Split: {}".format(self.what)
+
+
 def get_int(b):
     return int(codecs.encode(b, 'hex'), 16)
 
 
+def open_maybe_compressed_file(path):
+    """Return a file object that possibly decompresses 'path' on the fly.
+       Decompression occurs when argument `path` is a string and ends with '.gz' or '.xz'.
+    """
+    if type(path) != str: 
+        return path
+    if path.endswith('.gz'): 
+        return gzip.open(path, 'rb')
+    if path.endswith('.xz'):
+        return lzma.open(path, 'rb')
```

Comment:
maybe you can import `lzma` here instead?

Reply:
Sure.